extensions/amp-ad/0.1/amp-ad-batch-manager.js

```diff
@@ -0,0 +1,160 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {user} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+/** @const {number} The number of milliseconds between checks to see if the response has arrived */
+const AD_BATCH_TICK_LENGTH = 50;
+
+/** @const {number} The number of ticks to wait for a response from the ad server before timing out */
+const AD_BATCH_TIMEOUT = 100;
+
+/** @const {!string} Tag name for ad batch manager implementation. */
+export const TAG_AD_BATCH_MANAGER = 'amp-ad-batch-manager';
+
+/**
+  * Get a batch manager that will be used to batch together elements with the same data-url
+  * Note a side effect: the first ad with a given data-url to be processed will have its isBatchMaster_ variable set to true
+  * @param {Element} e The element which is to be batched
+  * @returns {AmpAdBatchManager}
+  */
+export function getBatchManager(e) {
+  // If this is our first imagead, create a map of responses for each value of 'data-url'
+  // This allows ads from multiple ad servers on the same page
+  if (!(e.win.hasOwnProperty('imageadBatchManagers'))) {
+    // Create an array of batch mangers. There will be one for each URL.
+    e.win.imageadBatchManagers = {};
+  }
+  if (!(e.url_ in e.win.imageadBatchManagers)) {
+    // This is the first time we've seen this URL, so this will be the master for this batch
+    e.isBatchMaster_ = true;
+    e.win.imageadBatchManagers[e.url_] = new AmpAdBatchManager(e);
+  } else {
+    e.isBatchMaster_ = false;
+  }
+  return e.win.imageadBatchManagers[e.url_];
+}
+
+export class AmpAdBatchManager {
+
+  /** @param {!Element} element */
+  constructor(element) {
+
+    /** @private {string} The base URL of the ad server */
+    this.url_ = element.url_;
+
+    /** @private {string} The full URL of the ad server, including any slot ids */
+    this.fullurl_ = this.url_;
+
+    /** @private {string} The number of timer ticks remaining before we give up waiting for an ad server response */
+    this.timeout_ = AD_BATCH_TIMEOUT;
+
+    /** @public {Object} responseData Map of JSON responses indexed by slot */
+    this.responseData = null;
+
+    // Scan the document for all elements with the same URL, and put them into the list to be batched
+    // @todo - the selector could be made a variable in order to generalise this to things other than imageads.
+    const elements = document.querySelectorAll('amp-ad[type=imagead]');
```

Comment:
I've changed the code so that the scan is only done once, although I've kept the one BatchManager per URL schema because it greatly simplifies the rest of the code. At the same time, I've made the selector a variable, allowing the batch manager to be used with any class of element.